Restructured  the layout of the current documentation

Roseline-Bassey opened the pull request:
This PR is an implementation of the proposed docs structure #306  for the Notary Project.

The documentation layout has been restructured in this order :

Project overview
Contributing.md
Notation >

- Quick-start guides.md
- Installation guides
- Tutorials
- How-to-guides
- Concepts ( this only contains the [Notation directory structure for system configuration](https://notaryproject.dev/docs/concepts/directory-structure/) now)
- CLI Reference
- Glossary.md ( some terminologies which I will define in my later PR have been added here)

Troubleshooting

Developer documentation ([Securely develop plugins for Notation](https://notaryproject.dev/docs/concepts/secure-plugin-development/), [Securely deploying Notation](https://notaryproject.dev/docs/concepts/secure-deployment/), [Experimental features](https://notaryproject.dev/docs/concepts/experimental-features/), and [Notary project specifications and requirements](https://notaryproject.dev/docs/concepts/specs-requirements/) have been moved to developer docs)

FAQ

Roseline-Bassey commented:
@zr-msft  #312 is the right one

Roseline-Bassey commented:
I am going to close this now.

zr-msft reviewed (changes requested):
@Roseline-Bassey 

Thank you for making this PR.

Overall i'm fine with the changes for now, but there a couple changes:

- Create `content/en/docs/notation/developer`
- Move `experimental-features.md`, `secure-deployment.md`, and `secure-plugin-development.md` under `content/en/docs/notation/developer`
- Rename `content/en/docs/developer-docs` -> `content/en/docs/developer` 
-

zr-msft commented:
fixes #306

Roseline-Bassey commented:
Hi, @zr-msft, thanks for the comment. I have updated the PR to incorporate your suggestions. Please kindly review.

zr-msft reviewed (dismissed):
Nice work. LGTM

Roseline-Bassey commented:
Thank you, @zr-msft

FeynmanZhou commented on content/en/docs/notation/developer/_index.md:
```diff
@@ -0,0 +1,5 @@
+---
+title: Developer documentation
```
Where there are two duplicated Developer documentation? 
![image](https://github.com/notaryproject/notaryproject.dev/assets/40452856/423be3c5-1f50-4d45-8e15-5d9b1263cb41)

zr-msft replied:
@FeynmanZhou  i asked for 2 different developer dirs:
- Developer docs for notation specifically (e.g. plugins)
- Developer docs for Notary (e.g. implementing the spec)

For the secure deploy and experimental features docs, I'm fine with moving them to notation/concepts too.

Roseline-Bassey replied:
I think both `Developer docs for notation` and
`Developer docs for Notary`  can be subfolders inside a single developer documentation directory. For example:

Developer documentation
-  Developer docs for notation (the secure plugin development, deployment, and experimental features docs would be here)
-  Developer docs for Notary (e.g. implementing the spec)

Please let me know what you think of this. @zr-msft & @FeynmanZhou

Roseline-Bassey replied:
Considering that the intended audience for `secure-deployment.md` and `experimental-features.md` are users, I could submit another PR to move the docs to `content.en/docs/notation/concepts` as @zr-msft said. 

To avoid duplicating developer documentation, the developer docs for notation and Notary can be grouped under a single directory. Here's an example of how it could be organized:

```
Troubleshooting
Developer documentation
- Notation Developer documentation
  - securely-develop-plugins.md
- Notary Developer documentation
  - implementing-specs.md
  ```      
This way, developers can access documentation for both Notation and Notary within a single directory.

Alternatively, we could have a single directory that includes documentation for both Notation and Notary like so:

```
Troubleshooting
Developer documentation
- securely-develop-plugins.md
- implementing-specs.md
```

please let me know what you think. @FeynmanZhou

zr-msft replied:
@Roseline-Bassey I think its best to keep the original idea of:
- `content/en/docs/notation/developer` 
- `content/en/docs/developer`

This allows us to version the docs for the notary project and notation independently

@Roseline-Bassey also, if you are making changes based on feedback in this PR, please add another commit to this PR or amend the current commit in this PR
eg:
> Considering that the intended audience for secure-deployment.md and experimental-features.md are users, I could submit another PR to move the docs to content.en/docs/notation/concepts ...

Roseline-Bassey replied:
I have submitted a PR to move `secure-deployment.md` and `experimental-features.md` to `content/en/docs/notation/concepts`

FeynmanZhou reviewed (changes requested):
PTAL

zr-msft commented:
@Roseline-Bassey i noticed you added an `en` directory. 

I'm fine with keeping this, and I'd like to add localization to the docs, but in the future, please include some explanation of additions like this in the description of the PR.

zr-msft commented:
@Roseline-Bassey you'll need to rebase your branch for this PR and resolve any conflicts before we can merge this

Roseline-Bassey commented:
Hi, @zr-msft 
I have rebased and force-pushed to my branch. On my local Git, the rebase output shows that it "Successfully rebased and updated refs/heads/refactoring-docs-branch."  There was no conflict on my local git after rebasing and before submitting the pull request but on this PR page GitHub says: 'This branch has conflicts that must be resolved" I am wondering if I pushed the PR wrongly and what can I do to resolve the conflict.

zr-msft commented:
> Hi, @zr-msft I have rebased and force-pushed to my branch. On my local Git, the rebase output shows that it "Successfully rebased and updated refs/heads/refactoring-docs-branch." There was no conflict on my local git after rebasing and before submitting the pull request but on this PR page GitHub says: 'This branch has conflicts that must be resolved" I am wondering if I pushed the PR wrongly and what can I do to resolve the conflict.

did you fetch the latest from upstream?

```
git fetch upstream
git rebase upstream/main
```

Roseline-Bassey commented:
On the `main` branch, I executed 
```git fetch upstream```

Then on my branch, I executed:
```git rebase main```

zr-msft commented:
> On the `main` branch, I executed `git fetch upstream`
> 
> Then on my branch, I executed: `git rebase main`

you need to run `git rebase upstream/main` on your branch. You ran the equivalent of `git rebase origin/main`

Roseline-Bassey commented:
Thanks, @zr-msft 
Everything is working fine now.

FeynmanZhou commented:
> This PR is an implementation of the proposed docs structure #306 for the Notary Project.
> 
> The documentation layout has been restructured in this order :
> 
> Project overview Contributing.md Notation >
> 
> * Quick-start guides.md
> * Installation guides
> * Tutorials
> * How-to-guides
> * Concepts ( this only contains the [Notation directory structure for system configuration](https://notaryproject.dev/docs/concepts/directory-structure/) now)
> * CLI Reference
> * Glossary.md ( some terminologies which I will define in my later PR have been added here)
> 
> Troubleshooting
> 
> Developer documentation ([Securely develop plugins for Notation](https://notaryproject.dev/docs/concepts/secure-plugin-development/), [Securely deploying Notation](https://notaryproject.dev/docs/concepts/secure-deployment/), [Experimental features](https://notaryproject.dev/docs/concepts/experimental-features/), and [Notary project specifications and requirements](https://notaryproject.dev/docs/concepts/specs-requirements/) have been moved to developer docs)
> 
> FAQ

@Roseline-Bassey What's the reason that we use `Notation` as the top-level name of this chapter? 
@zr-msft Do you feel this is a reasonable directory structure?

Roseline-Bassey commented:
`Notation` is used as a top-level name to serve as a central directory for all content related to notation for better organization of content and versioning. @FeynmanZhou

zr-msft commented:
@FeynmanZhou the content at the root level is related to the Notary Project in general. Content in the `notation` dir is specific to notation. There could be a future state where this another client that does notary signatures that would be a peer to notation and have a dir at the same level. Content at the root level for the Notary Project would apply to both clients.

@toddysm please let me know if you disagree

FeynmanZhou commented:
> @FeynmanZhou the content at the root level is related to the Notary Project in general. Content in the `notation` dir is specific to notation. There could be a future state where this another client that does notary signatures that would be a peer to notation and have a dir at the same level. Content at the root level for the Notary Project would apply to both clients.
> 
> @toddysm please let me know if you disagree

There is no plan for developing another client as a peer based on the Notary Project specifications in the near future. I would suggest focusing on a certain thing (currently there is Notation only) and designing the documentation directory to be more straightforward. In case another client is planned in the future, we could refactor the documentation directory accordingly.

zr-msft commented:
> > @FeynmanZhou the content at the root level is related to the Notary Project in general. Content in the `notation` dir is specific to notation. There could be a future state where this another client that does notary signatures that would be a peer to notation and have a dir at the same level. Content at the root level for the Notary Project would apply to both clients.
> > @toddysm please let me know if you disagree
> 
> There is no plan for developing another client as a peer based on the Notary Project specifications in the near future. I would suggest focusing on a certain thing (currently there is Notation only) and designing the documentation directory to be more straightforward. In case another client is planned in the future, we could refactor the documentation directory accordingly.
> 
> I will consult other maintainers @iamsamirzon @toddysm @yizha1 about this new documentation directory change.

@FeynmanZhou Having Notation content clearly separated from Notary Project content helps distinguish the difference between the two. As a side benefit, it also makes it easier for other clients in the future.

Having a more streamlined content directory structure would solve some complexity problems, but it introduces others. For example, versioning the content for new Notation releases. If we create a version 1.1 of the docs and we tightly couple Notary Project content with Notation content, does that mean both Notary Project and Notation are at 1.1? If its only for Notation, how can we show that if content is mixed together?

yizha1 commented:
My comments:

The first entry includes documents for Notary Project, like overview, specifications (current it is under Concept), contributing, FAQ for Notary Project and folders for sub-projects, for example Notation at this moment.

- Project overview (Notary Project Overview)
- Notation (Notation CLI, Libraries and Plugins)
  - Quick-start
  - Installation guides
  - Tutorials
  - How-to-guides
  - Concepts (keep only Notation specific)
  - CLI Reference
  - Troubleshooting
  - Development document
  - FAQ (for Notation)
- Specification (Notary Project specifications)
- Contributing
- FAQ (for Notary Project)

zr-msft commented:
@yizha1 I'm fine updating the order

@FeynmanZhou i understand your approach about user types and progression, but I don't think that works for organizing the docs. Specifically, it causes the following issues:
- It tightly couples The Notary Project with Notation which i thought we were trying to avoid
- It introduces versioning problems.

I'll reiterate my concerns from my [previous comment](https://github.com/notaryproject/notaryproject.dev/pull/312#issuecomment-1648465861): In this tightly-coupled structure, what happens when Notation releases a new version, but the Notary Project is the same version? Does the whole project assume the version of Notation? What happens in the future if there is another client under the Notary Project that needs to be documented with its own version independent of Notation?

FeynmanZhou commented:
@zr-msft Thanks. I understand your major concern, which is about the versioning problems if we mix different sub-projects in the documentation.  

Actually, the term `Notary Project` is the brand name and refers to all sub-projects as described in https://github.com/notaryproject/.github/issues/35 (TLDR, See summary below). So `Notary Project` will not be versioned separately. [All sub-projects in the Notary Project organization](https://github.com/notaryproject) will have their own versioning, but those sub-projects should be documented on the website. Most of the sub-projects are created for Notation.

> The name of the GitHub organization is "Notary Project". When used this term has an all-encompassing meaning and will refer to the GitHub organization, the community, all specifications, and all the repositories under the organization including tools that are released by the community.

To resolve your questions below, I would suggest the current directory structure focuses on Notation at this stage since Notation is the only core sub-project. Most of the documentation content is also created for Notation. If there is another client tool in the future, we can design a separate documentation entry and provide a menu accordingly. 

> In this tightly-coupled structure, what happens when Notation releases a new version, but the Notary Project is the same version? Does the whole project assume the version of Notation? What happens in the future if there is another client under the Notary Project that needs to be documented with its own version independent of Notation?

FeynmanZhou commented:
> Are you all planning on releasing all the sub-projects under The Notary Project as a system release with the same version? 

All sub-projects will have their separate versioning not only notation, notation-go, notation-core-go. For Go library docs, they are maintained on [Go doc](https://pkg.go.dev/github.com/notaryproject/notation-go@main?tab=versions) so we don't have too much additional content needs to be added to the website. For other development guides, I suggest maintaining the latest version only since it doesn't change along with different Notation versions generally. We can state this in the development docs.

> `https://notaryproject.dev/docs/1.1/developer/go-library` is the doc for developing against the go library when version 1.1 of notation is released. The version in the URL has no connection to the version of the go-library, but looks like it does. It's not obvious that the version is only tied to notation.

We will maintain only two versions of all sub-projects including "N" and "N-1" as described in the [RELEASE MANAGEMENT](https://github.com/notaryproject/notation/blob/main/RELEASE_MANAGEMENT.md#supported-releases). So I would expect the default URL without a version number that implies the latest version, such as `https://notaryproject.dev/docs/developer/go-library`. Accordingly, users only specify a version when they want to explore the docs in an N-1 version.

> https://notaryproject.dev/docs/1.1/overview/ is the doc giving an overview of The Notary Project when version 1.1 of notation is released. The content has no relation to the version in the URL, but it looks like it does. It could easily be confused as the doc giving an overview of The Notary Project at version 1.1 of The Notary Project, which is not true.

This is a bit confused indeed since Notation and Notary Project are tightly coupled in the project overview doc. The potential solution could be focused on the `Notation` but briefly introduce `Notary Project` and link to [.github README](https://github.com/notaryproject/.github). There will be a comprehensive introduction to the overall Notary Project. We can also state the `Notary Project` doesn't have a separate versioning accordingly.

FeynmanZhou commented:
Let's discuss the version control and new directory structure in the Notary Project community meeting later : )

toddysm commented:
After the discussion in the Notary Project Community meeting, I like the proposal from Sajay to have a project (system release) version that wil be used for the documentation. The system release is a combination of several things (spec, CLI, libraries) that can have different versions - we can describe those in the system release overview. For the current release, we can either avoid the version in the documentation URL or use 1.0 - IMHO for this release doesn't matter too much because everything will be 1.0.0. In the future we should have this clearly described in the Overview.

zr-msft commented:
If we agree to a system release per the message from @toddysm, I propose we update the directory structure refactor to the following:

```
docs/
    notary-project-overview.md
    notary-project-concepts.md
    faq.md
    user-guides/
        quickstart.md
        installation-guides/
            cli.md
            uninstall.md
        tutorials/
            trust-policy.md
        how-to-guides/
            manage-signatures.md
            ...
            registry-authentication.md
        notation-cli-reference/
            notation_certificate-add.md
            ...
            notation.md
        troubleshooting.md
    developer-guides/
        notation-plugin-developement.md
        notation-core-go-development.md
        notation-go-development.md
    contribute.md
    notary-specification.md
```

This incorporates many points of feedback and suggestions from myself as well as @FeynmanZhou @Roseline-Bassey and @yizha1 

@toddysm @FeynmanZhou @Roseline-Bassey @yizha1  @sajayantony @iamsamirzon what do you think?

FeynmanZhou commented:
Agree with the proposal from @toddysm @zr-msft  and the latest directory structure suggested by @Roseline-Bassey .

Just one note, for those proposed docs without content in chapters like `developer-guides`, please avoid creating empty placeholder `.md` files in this PR. We can send follow-up PRs to those docs until we have certain content. 
-

yizha1 commented:
+1 to @toddysm 's [proposal ](https://github.com/notaryproject/notaryproject.dev/pull/312#issuecomment-1659402601), and directory structure details shared by @zr-msft @Roseline-Bassey .

Roseline-Bassey commented:
I'm currently working on updating this PR now. @zr-msft

iamsamirzon commented:
Where can I see the latest preview of the .dev website?

Roseline-Bassey commented:
Here's the latest preview of the website: https://deploy-preview-312--notarydev.netlify.app/ 
cc: @iamsamirzon

zr-msft commented:
@Roseline-Bassey  can you rebase again when you have a chance and resolve the conflicts?

Roseline-Bassey commented:
I have rebased this branch @zr-msft

iamsamirzon commented:
I am unable to leave inline comments, so summarizing here some feedback on FAQs. Originally this section was meant to refer to only things related to the new specification and its implementation Notation, but recently we added FAQs about the notary client, notation client, and the overall notary project. Consequently every prior use of "Notary" has to be revisited on this page and website , so we can either change it to "Notary Project", "Notary Specification", or "Notation" client. For instance refer the 4 FAQs under the sub-heading  https://deploy-preview-312--notarydev.netlify.app/docs/faq/#trust-store-and-trust-policy  . They should all be referring to Notation Specification

iamsamirzon commented on content/en/docs/user-guides/how-to/oci-image-layout.md:
```diff
@@ -1,7 +1,7 @@
+weight: 5
 ---
 
 OCI image layout is a directory structure that contains files and folders that refer to an OCI image. OCI image layout is defined in [OCI image spec](https://github.com/opencontainers/image-spec/blob/v1.1.0-rc2/image-layout.md).
```
Adding a comment in this line, because I could not find a way to leave comment for the next line which needs changing

~~When you sign an image after it is pushed to a registry, it is possible that you are not signing the correct image. You also may not be certain that the image hasn't been altered.~~ Using OCI image layout, you can sign images on local disk before pushing them to the remote registry. ~~Signing an image before you push it a registry provides helps ensure you are signing the correct image, and consumers of that image can verify that same image.~~

When you sign an image after it is pushed to a registry, it is possible that you are not signing the correct image. You also may not be certain that the image hasn't been altered. Using OCI image layout, you can sign images on local disk before pushing them to the remote registry. Signing an image before you push it a registry provides helps ensure you are signing the correct image, and consumers of that image can verify that same image.

iamsamirzon commented on content/en/docs/faq.md:
```diff
@@ -2,7 +2,7 @@
+weight: 7
 ---
 
 ## What registries are compatible with Notary?
```
```suggestion
## What registries are compatible with Notation?
```
Also need to make the same change in the next line "The following registries are compatible with ~~Notary~~ Notation for OCI artifact signing and verification:

iamsamirzon commented on content/en/docs/faq.md:
```diff
@@ -2,7 +2,7 @@
 title: "Frequently asked questions"
 description: "Frequently asked questions about Notary"
```
There are multiple other instances on this page, where "Notary" has to either change to "Notation" or "Notary Project"

FeynmanZhou commented on content/en/docs/notary-specification.md:
```diff
@@ -2,7 +2,7 @@
 title: "Notary project specifications and requirements"
 description: "Notary project specifications and requirements"
```
Notary Project

FeynmanZhou commented on content/en/docs/notary-specification.md:
```diff
@@ -2,7 +2,7 @@
 
-More information on the specifications, requirements, and other details about the Notary project can be found in the [Notary repository](https://github.com/notaryproject/notaryproject/tree/main).
\ No newline at end of file
+More information on the specifications, requirements, and other details about the Notary project can be found in the [Specification repository](https://github.com/notaryproject/specifications/tree/main).
```
Notary Project

FeynmanZhou commented on content/en/docs/faq.md:
```diff
@@ -56,12 +56,12 @@ This has implication such as an end user with CA issued certificate can masquera
 ## Trust store and trust policy
 
-**Q: Does Notary supports `n` out of `m` signatures verification requirement?**
+**Q: Does Notation Specification supports `n` out of `m` signatures verification requirement?**
```
```suggestion
**Q: Does Notary Project specification supports `n` out of `m` signatures verification requirement?**
```

FeynmanZhou commented on content/en/docs/faq.md:
```diff
@@ -56,12 +56,12 @@ This has implication such as an end user with CA issued certificate can masquera
+**Q: Does Notation Specification supports `n` out of `m` signatures verification requirement?**
 
-**A:** Notary doesn't support n out m signature requirement verification scheme.
+**A:** Notation Specification doesn't support `n` out of `m` signature requirement verification scheme.
```
```suggestion
**A:** Notary Project specification doesn't support `n` out of `m` signature requirement verification scheme.
```

FeynmanZhou commented on content/en/docs/faq.md:
```diff
@@ -56,12 +56,12 @@ This has implication such as an end user with CA issued certificate can masquera
 Signature verification workflow succeeds if verification succeeds for at least one signature.
 
-**Q: Does Notary support overriding of revocation endpoints to support signature verification in disconnected environments?**
+**Q: Does Notation Specification support overriding of revocation endpoints to support signature verification in disconnected environments?**
```
```suggestion
**Q: Does Notation support overriding of revocation endpoints to support signature verification in disconnected environments?**
```

FeynmanZhou commented on content/en/docs/user-guides/_index.md:
```diff
@@ -0,0 +1,5 @@
+---
+title: User-guides
```
Should it be `User guides`?

Roseline-Bassey replied:
Yes, `User guides` is commonly used.

FeynmanZhou commented on content/en/docs/faq.md:
```diff
@@ -18,25 +18,25 @@ The following registries are compatible with Notary for artifact signing and ver
 **Q: Why JWT `exp` and `iat` claims are not used?**
 
-**A:** Unlike JWT which always contains a JSON payload, Notary envelope can support payloads other than JSON, like binary. Reusing the JWT payload structure and claims, limits the Notary JWS envelope to only support JSON payload, which is undesirable. Also, reusing JWT claims requires following same claim semantics as defined in JWT specifications. The [`exp`](https://datatracker.ietf.org/doc/htm ...
+**A:** Unlike JWT which always contains a JSON payload, Notation envelope can support payloads other than JSON, like binary. Reusing the JWT payload structure and claims, limits the Notation JWS envelope to only support JSON payload, which is undesirable. Also, reusing JWT claims requires following same claim semantics as defined in JWT specifications. The [`exp`](https://datatracker.ietf.org/doc ...
```
```suggestion
**A:** Unlike JWT which always contains a JSON payload, the Notary Project signature envelope can support payloads other than JSON, like binary. Reusing the JWT payload structure and claims, limits the Notary Project signature JWS envelope to only support JSON payload, which is undesirable. Also, reusing JWT claims requires following the same claim semantics as defined in JWT specifications. The [`exp`](https://datatracker.ietf.org/doc/html/rfc7519#section-4.1.4) claim requires that the verifier MUST reject the signature if the current time equals or is greater than `exp`, where as Notation allows verification policy to define how expiry is handled.
```

FeynmanZhou reviewed (commented):
PTAL

FeynmanZhou commented:
Hi @Roseline-Bassey , the updated version looks good to me. Thanks for your patience. Could you ps resolve the conflict?

FeynmanZhou commented:
@zr-msft @Roseline-Bassey  I feel this article could be moved to the chapter "User guides" since only Notation users may take care of this configuration. What's your opinion?
![image](https://github.com/notaryproject/notaryproject.dev/assets/40452856/6c5b08fa-8efd-4018-849b-6fb9c1614b7a)

Roseline-Bassey commented:
Yes, I agree. The article can be included in the how-to guides under the User Guides. @FeynmanZhou

FeynmanZhou reviewed (approved):
LGTM

zr-msft reviewed (approved):
LGTM
